### PR TITLE
feat: add slide_from_bottom & fade_from_bottom animations

### LINKED ISSUE #851: native-stack animation does not fade in from the bottom on Android
## Description

When using `createNativeStackNavigator`, the animation is strange on Android.
Specifically, the animation scale from center(zoom-out-from-center), which is not a standard OS animation.

The official README for [createNativeStackNavigator](https://github.com/software-mansion/react-native-screens/tree/a46aaab4b594079cdce36ff0115f472f5249c506/createNativeStackNavigator)
 says to expect the standard OS animation of fading in from the bottom.
> By default the stack navigator is configured to have the familiar iOS and Android look & feel: new screens slide in from the right on iOS, fade in from the bottom on Android. On iOS, the stack navigator can also be configured to a modal style where screens slide in from the bottom.

The current animation is different from the expected behavior.

## Steps To Reproduce

1. Build and run the Example in this repository on Android. https://github.com/software-mansion/react-native-screens/tree/a46aaab4b594079cdce36ff0115f472f5249c506/Example
2. Select menu the `Native stack bindings for RNN`
3. Select `PUSH` or `MODAL` to cause a navigation animation.

### Expected behavior

fade in animation from the bottom on Android

### Actual behavior

scale-from-center(zoom-out-from-center) animation on Android

## Snack or minimal code example

The Example of this repository.

https://github.com/software-mansion/react-native-screens/tree/a46aaab4b594079cdce36ff0115f472f5249c506/Example

## Package versions

- React: 16.13.1
- React Native: 0.63.0
- React Native Screens: 2.18.1

### PR DESCRIPTION
**Pull request by kacperkapusciak**
## Description
PR originally authored by @itsjgf in https://github.com/software-mansion/react-native-screens/pull/662
Add support for `slide_from_bottom` and `fade_from_bottom` animations on both platforms.
`fade_from_bottom` animation mimics default Android Nougat animation.

Fixes #851, #651

## Changes

- Added slide_from_bottom animation
- Added fade_from_bottom animation
- Added example
- Updated types
- Updated documentation

## Screenshots / GIFs

`slide_from_bottom` on iOS

https://user-images.githubusercontent.com/39658211/123950477-04017380-d9a4-11eb-80ba-82c3804fffac.mp4

`slide_from_bottom` on Android

https://user-images.githubusercontent.com/39658211/123950560-1c718e00-d9a4-11eb-9176-2bdaab235f0f.mp4

`fade_from_bottom` on iOS

https://user-images.githubusercontent.com/39658211/123950617-2bf0d700-d9a4-11eb-820e-2c324d597ff8.mp4

`fade_from_bottom` on Android

https://user-images.githubusercontent.com/39658211/123950690-3ca14d00-d9a4-11eb-8195-d270e0eac5ea.mp4

## Checklist

- [x] Included code example that can be used to test this change
- [x] Updated TS types
- [x] Updated documentation: 
  - [x] https://github.com/software-mansion/react-native-screens/blob/master/README.md
  - [x] https://github.com/software-mansion/react-native-screens/blob/master/native-stack/README.md
  - [x] https://github.com/software-mansion/react-native-screens/blob/master/createNativeStackNavigator/README.md
  - [x] https://github.com/software-mansion/react-native-screens/blob/master/src/types.tsx
  - [x] https://github.com/software-mansion/react-native-screens/blob/master/src/native-stack/types.tsx
- [x] Ensured that CI passes

### CODE REVIEW

**Comment by jp928** [comment]
Knee to have this. Thanks @kacperkapusciak

File: android/src/main/java/com/swmansion/rnscreens/ScreenStack.java [hunk 1]
```diff
@@ -341,22 +362,29 @@ private void setupBackHandlerIfNeeded(ScreenStackFragment topScreen) {
...
+          .beginTransaction()
+          .show(topScreen)
+          .addToBackStack(BACK_STACK_TAG)
+          .setPrimaryNavigationFragment(topScreen)
+          .commitAllowingStateLoss();
       mFragmentManager.addOnBackStackChangedListener(mBackStackListener);
     }
   }
 
   private static boolean isCustomAnimation(Screen.StackAnimation stackAnimation) {
```
Comment (kacperkapusciak) [inline android/src/main/java/com/swmansion/rnscreens/ScreenStack.java:377]: After you've inverted the logic here it should no longer be named `isCustomAnimation`.

File: android/src/main/res/anim/rns_identity.xml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (kacperkapusciak) [inline android/src/main/res/anim/rns_identity.xml:1]: I suggest renaming this file to a more human name. I'm aware that `identity` is the true 'correct' name for this but I'm keen on calling it something less smart but more approachable e.g. `no_animation`, `do_nothing`

File: android/src/main/java/com/swmansion/rnscreens/ScreenStack.java [hunk 3]
```diff
@@ -240,6 +256,13 @@ protected void performUpdate() {
     }
     // animation logic end
 
+    if (mDismissed.isEmpty() && mTopScreen != null && isSlideFromBottom(mTopScreen) && visibleBottom == null) {
+      // When going forward in `slide_from_bottom` transition, we want the previous screen, which is mTopScreen now,
+      // to stay visible during transition, we remove it after the transition ended
```
Comment (kacperkapusciak) [inline android/src/main/java/com/swmansion/rnscreens/ScreenStack.java:261]: I'd love to see here "we remove it in XYZ method after...". That would pinpoint where exactly it is being removed and then everything would be clear.

**Review by kacperkapusciak — COMMENTED** [review]
Just one small issue within types

File: android/src/main/java/com/swmansion/rnscreens/ScreenStack.java [hunk 4]
```diff
@@ -367,4 +408,81 @@ private void setupBackHandlerIfNeeded(ScreenStackFragment topScreen) {
...
+    if (drawingOps.size() < previousChildrenCount) {
+      reverseLastTwoChildren = false;
+    }
+    previousChildrenCount = drawingOps.size();
+
+    if (reverseLastTwoChildren && drawingOps.size() >= 2) {
+      Collections.swap(drawingOps, drawingOps.size() - 1, drawingOps.size() - 2);
+    }
+
+    while (!drawingOps.isEmpty()) {
```
Comment (kmagiera) [inline android/src/main/java/com/swmansion/rnscreens/ScreenStack.java:447]: i wonder why cant we just have a simple loop over items from drawingOps here. this isnt very crutial i believe, but still seems a bit obscure and strikes me when such iteration is done in n^2 while simple O(n) loop would suffice

File: android/src/main/java/com/swmansion/rnscreens/ScreenStackFragment.java [hunk 5]
```diff
@@ -232,10 +232,21 @@ public NotifyingCoordinatorLayout(@NonNull Context context, ScreenFragment fragm
...
       // exit transitions so we use this hack.
-      AnimationSet set = new AnimationSet(true);
-      set.addAnimation(animation);
-      set.setAnimationListener(mAnimationListener);
-      super.startAnimation(set);
+      // We also have some animations that are an AnimationSet, so we don't wrap them
+      // in another set since it causes some visual glitches when going forward.
+      // We also set the listener only when going forward, since when going back
+      // there is already a listener for dismiss action, which would be overridden
+      // and also the hack is not necessary when going back since the lifecycle methods
```
Comment (kmagiera) [inline android/src/main/java/com/swmansion/rnscreens/ScreenStackFragment.java:239]: the word hack has some bad connotations and id prefer not to use it when other words seem more appropriate

File: android/src/main/res/anim/rns_fade_from_bottom.xml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+<!--http://aosp.opersys.com/xref/android-7.1.2_r37/xref/frameworks/base/core/res/res/anim/activity_open_enter.xml-->
+<set xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shareInterpolator="false">
+    <alpha android:fromAlpha="0.0" android:toAlpha="1.0"
+        android:interpolator="@android:interpolator/decelerate_quint"
+        android:duration="210"/>
+    <translate android:fromYDelta="8%" android:toYDelta="0"
+        android:interpolator="@android:interpolator/decelerate_quint"
+        android:duration="350"/> <!--we use rns_no_animation_350.xml as the other animation for
+        this transition in oder for them to end at the same time. Remember to change
```
Comment (kmagiera) [inline android/src/main/res/anim/rns_fade_from_bottom.xml:12]: something is wrong with this comment

File: android/src/main/res/anim/rns_fade_to_bottom.xml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+<set xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shareInterpolator="false">
+    <alpha android:fromAlpha="1.0" android:toAlpha="0.0"
+        android:interpolator="@android:interpolator/linear"
+        android:startOffset="100"
+        android:duration="150"/>
+    <translate android:fromYDelta="0%" android:toYDelta="8%"
+        android:interpolator="@android:interpolator/accelerate_quint"
+        android:duration="250"/>  <!--we use rns_no_animation_250.xml as the other animation for
+        this transition in oder for them to end at the same time. Remember to change
```
Comment (kmagiera) [inline android/src/main/res/anim/rns_fade_to_bottom.xml:13]: same here and perhaps in oder related places

File: ios/RNSScreenStack.m [hunk 8]
```diff
@@ -520,10 +520,8 @@ - (void)dismissOnReload
     screen = (RNSScreenView *)fromVC.view;
   }
   if (screen != nil &&
-      (screen.stackAnimation == RNSScreenStackAnimationFade ||
-       screen.stackAnimation == RNSScreenStackAnimationSimplePush ||
-       screen.stackAnimation == RNSScreenStackAnimationNone ||
-       screen.stackAnimation == RNSScreenStackAnimationSlideFromBottom)) {
+      (screen.stackAnimation != RNSScreenStackAnimationFlip && screen.stackAnimation != RNSScreenStackAnimationNone &&
```
Comment (kmagiera) [inline ios/RNSScreenStack.m:523]: why is flip animation excluded here?

File: ios/RNSScreenStackAnimator.m [hunk 9]
```diff
@@ -186,4 +190,63 @@ - (void)animateSlideFromBottomWithTransitionContext:(id<UIViewControllerContextT
...
+                     }
+                     completion:nil];
+    [UIView animateWithDuration:0.15
+        delay:0.1
+        options:UIViewAnimationOptionCurveLinear
+        animations:^{
+          fromViewController.view.alpha = 0.0;
+        }
+        completion:^(BOOL finished) {
+          [transitionContext completeTransition:![transitionContext transitionWasCancelled]];
```
Comment (kmagiera) [inline ios/RNSScreenStackAnimator.m:247]: this aniation is shorter while we call complete handler in it. shouldnt that completion call be after the last animation is complete?

File: android/src/main/java/com/swmansion/rnscreens/ScreenStack.java [hunk 10]
```diff
@@ -367,4 +408,81 @@ private void setupBackHandlerIfNeeded(ScreenStackFragment topScreen) {
...
   }
+
+  // below methods are taken from
+  // https://github.com/airbnb/native-navigation/blob/9cf50bf9b751b40778f473f3b19fcfe2c4d40599/lib/android/src/main/java/com/airbnb/android/react/navigation/ScreenCoordinatorLayout.java#L43
+  // and are used to swap the order of drawing views when navigating forward with the transitions
+  // that are making transitioning fragments appear one on another. See more info in the comment to
+  // the linked class.
+  @Override
+  public void removeView(final View view) {
+    if (isDetachingCurrentScreen) {
```
Comment (kmagiera) [inline android/src/main/java/com/swmansion/rnscreens/ScreenStack.java:419]: i dont understand the logic on when this property should be reset back to `true` -- can this be better explained somwhere in the comments

**Review by kmagiera — APPROVED** [review]
left a few comments. there is othing super major with this pr and looks good overall.

File: ios/RNSScreenStack.m (hunk 8 above)
Comment (WoLewicki) [reply ios/RNSScreenStack.m:523]: I changed it to exclude the animations handled by the system in order not to add new items to `if` clause when adding new animations. Also, `RNSScreenStackAnimationNone` should not be in here, will change it.

File: android/src/main/java/com/swmansion/rnscreens/ScreenStack.java (hunk 10 above)
Comment (WoLewicki) [reply android/src/main/java/com/swmansion/rnscreens/ScreenStack.java:419]: I will add a small comment here too, but the main one will be in the only place in code when we change it to `true`.

**Pull request merged**